Sources/LanguageServerProtocol/SupportTypes/RegistrationOptions.swift

```diff
@@ -26,21 +37,57 @@ public struct TextDocumentRegistrationOptions: RegistrationOptions {
     self.documentSelector = documentSelector
   }
 
-  public init?(fromLSPDictionary dictionary: [String : LSPAny]) {
-    guard let selectorValue = dictionary[CodingKeys.documentSelector.stringValue] else {
-      self.documentSelector = nil
-      return
+  public func encodeIntoLSPAny(dict: inout [String: LSPAny]) {
+    guard let documentSelector = documentSelector else { return }
+    dict["documentSelector"] = documentSelector.encodeToLSPAny()
+  }
+
+  public static func == (lhs: TextDocumentRegistrationOptions, rhs: TextDocumentRegistrationOptions) -> Bool {
+    return lhs.documentSelector == rhs.documentSelector
+  }
+
+  public func hash(into hasher: inout Hasher) {
+    documentSelector?.hash(into: &hasher)
+  }
+}
+
+/// Protocol for a type which structurally represents`TextDocumentRegistrationOptions`.
+public protocol TextDocumentRegistrationOptionsProtocol {
+  var textDocumentRegistrationOptions: TextDocumentRegistrationOptions {get}
+}
+
+/// Code completiion registration options.
+public struct CompletionRegistrationOptions: RegistrationOptions, TextDocumentRegistrationOptionsProtocol, Hashable {
+  public var textDocumentRegistrationOptions: TextDocumentRegistrationOptions
+  public var completionOptions: CompletionOptions
+
+  public init(documentSelector: DocumentSelector? = nil, completionOptions: CompletionOptions) {
+    self.textDocumentRegistrationOptions =
+        TextDocumentRegistrationOptions(documentSelector: documentSelector)
+    self.completionOptions = completionOptions
+  }
+
+  public func encodeIntoLSPAny(dict: inout [String: LSPAny]) {
+    textDocumentRegistrationOptions.encodeIntoLSPAny(dict: &dict)
+
+    if let resolveProvider = completionOptions.resolveProvider {
+      dict["resolveProvider"] = .bool(resolveProvider)
     }
-    guard case .dictionary(let selectorDictionary) = selectorValue else { return nil }
-    guard let documentSelector = DocumentSelector(fromLSPDictionary: selectorDictionary) else {
-      return nil
+    if let triggerCharacters = completionOptions.triggerCharacters {
+      dict["triggerCharacters"] = encode(strings: triggerCharacters)
+    }
+    if let allCommitCharacters = completionOptions.allCommitCharacters {
+      dict["allCommitCharacters"] = encode(strings: allCommitCharacters)
     }
-    self.documentSelector = documentSelector
   }
 
-  public func encodeToLSPAny() -> LSPAny {
-    guard let documentSelector = documentSelector else { return .dictionary([:]) }
-    return .dictionary([CodingKeys.documentSelector.stringValue: documentSelector.encodeToLSPAny()])
+  public static func == (lhs: CompletionRegistrationOptions, rhs: CompletionRegistrationOptions) -> Bool {
```

Comment:
Similarly, this should be able to use the synthesized `==` and `hash(into:)` now, right?

Reply:
Yup done